### PR TITLE
feat: allow disabling focusout event listeners in FloatingFocusManager

### LINKED ISSUE #2017: FloatingFocusManager focusout handler - closes when clicking another button, does not close on outside click
I created a small reproduction in a branch in my fork: https://github.com/mihkeleidast/floating-ui/commit/c5b92d4b029d8902e5b5a3178cb4cb4b3072ad37

Ran across an inconsistency in how the focusout handler works. Two different scenarios at play here:

Scenario 1
1. Open Popover from the button
2. Click outside of the Popover, observe that it does not close
3. Click the "Second button", observe that the Popover remains open

Scenario 2
1. Open Popover from the button
2. Click the "Second button" observe that the Popover closes

As far as I debugged, the focusout handler fires in both cases when clicking outside of the floating element, but in scenario 1, when clicking the background, `relatedTarget` is `null`, so it does not fire the open change event.

It is kind of an edge case, but would be good if the behavior was consistent, to know what I should expect, as from the users perspective a random click outside should not change anything. 

Some additional info: in my real usecase, I do not have such Popovers, rather it is a drawer-like component (ex https://shoelace.style/components/drawer) that can be configured to not render on top of other content, rather it would render aside the main page. In this context, I have floating-ui configured to not close the floating element on outside click, but this issue means it still does in some cases. I can create a more realistic sandbox, if necessary, though I think the above describes the issue in enough detail.

### PR DESCRIPTION
**Pull request by mihkeleidast**
Probably resolves #2017.

I added a somewhat realistic Drawer component to the playground as well, so the necessary behavior could be tested against and understood more easily. I don't really want to add too much of my specific components here, if they seem too much, so let me know if I should remove the example.

Additional comments:
- The added example can also be used to demo & test https://github.com/floating-ui/floating-ui/issues/2021 very well
- I kind of assumed that disabling the focusOut behavior here would keep the floating element open once I tab out of it, but it seems that there is another close hook in the focus guard that is rendered by FloatingPortal. Not too big of an issue at the moment, but maybe there needs to be another prop on FloatingPortal as well?

### CODE REVIEW

**Comment by atomiks** [comment]
I think it should be called `closeOnFocusOut`, it also only applies to non-modal focus management but the modal prop has some side-effects that should be maintained here?

**Comment by mihkeleidast** [comment]
> I think it should be called closeOnFocusOut

Renamed 👍 

> it also only applies to non-modal focus management but the modal prop has some side-effects that should be maintained here?

Did not quite get this, what sort of side effects of the modal prop should be mentioned here?

**Comment by atomiks** [comment]
This prevents screen readers from accessing outside content. Usually they won't be using `modal={true}` and `closeOnFocusOut={true}` together, but in case they are it should be guarded.

https://github.com/floating-ui/floating-ui/blob/bc3fb7cb262340187c1847ce2c9d1a0e7877c246/packages/react-dom-interactions/src/FloatingFocusManager.tsx#L255-L262

Btw, the docs will update before release, so usually have to do them separately

**Comment by mihkeleidast** [comment]
> Usually they won't be using modal={true} and closeOnFocusOut={true} together, but in case they are it should be guarded.

I assume you meant `modal={true} closeOnFocusOut={false}`, as both true is the default? Looking at the hideOthers/cleanup logic, why is it in this focusout effect at all, shouldn't it be in a separate effect?

I'll move the docs to a separate PR 👍

**Comment by mihkeleidast** [comment]
Docs are now in #2023. I also moved the hideOthers/cleanup logic to a separate effect, I don't think it should be disabled together with the rest of closeOnFocusOut. Not sure about all the hook deps, though - there were previously some that weren't used in the effect, not sure if they were deliberately set there for some side effects or not, you probably now best.

File: packages/react-dom-interactions/package.json [hunk 1]
```diff
@@ -85,7 +85,9 @@
     "react": "^18.0.0",
     "react-dom": "^18.0.0",
     "react-merge-refs": "^1.1.0",
+    "react-responsive": "^9.0.2",
```
Comment (atomiks) [inline packages/react-dom-interactions/package.json:88]: Why's this here?

Comment (mihkeleidast) [reply packages/react-dom-interactions/package.json:88]: I added it to properly showcase a component that changes the floating hook & component config based on screen size. Could also remove it and write a one-off resize handler for this, if adding it is too much.

**Pull request merged**